libraries/lib-numeric-formats/tests/NumericConverterTests.cpp

```diff
@@ -0,0 +1,157 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  NumericConverterTests.cpp
+
+  Dmitry Vedenko
+
+**********************************************************************/
+
+#include <catch2/catch.hpp>
+
+#include "formatters/ParsedNumericConverterFormatter.h"
+#include "formatters/BeatsNumericConverterFormatter.h"
+#include "NumericConverterFormatterContext.h"
+
+#include "Project.h"
+#include "ProjectRate.h"
+#include "ProjectTimeSignature.h"
+
+#include "MockedPrefs.h"
+#include "MockedAudio.h"
+
+TEST_CASE("ParsedNumericConverterFormatter", "")
+{
+   auto context = FormatterContext::SampleRateContext(44100.0);
+   
+   auto hhmmssFormatter = CreateParsedNumericConverterFormatter(
+      context, NumericConverterType_TIME, Verbatim("0100 h 060 m 060 s"));
+
+   REQUIRE(
+      hhmmssFormatter->ValueToString(0.0, false).valueString ==
+      "00 h 00 m 00 s");
+
+   REQUIRE(
+      hhmmssFormatter->ValueToString(30.0, false).valueString ==
+      "00 h 00 m 30 s");
+
+   REQUIRE(
+      hhmmssFormatter->ValueToString(60.0, false).valueString ==
+      "00 h 01 m 00 s");
+
+   REQUIRE(
+      hhmmssFormatter->ValueToString(60.0, false)
+         .fieldValueStrings.size() == 3);
+
+   REQUIRE(
+      hhmmssFormatter->ValueToString(60.0, false).fieldValueStrings[0] ==
+      "00");
+
+   REQUIRE(
+      hhmmssFormatter->ValueToString(60.0, false).fieldValueStrings[1] ==
+      "01");
+
+   REQUIRE(
+      hhmmssFormatter->ValueToString(60.0, false).fieldValueStrings[2] ==
+      "00");
+
+   REQUIRE(
+      hhmmssFormatter->StringToValue("foobar").has_value() == false);
+   REQUIRE(
+      hhmmssFormatter->StringToValue("01 h 30 m 15 s").has_value() == true);
+   REQUIRE(
+      *hhmmssFormatter->StringToValue("01 h 30 m 15 s") ==
+      Approx(60 * 60 + 30 * 60 + 15));
+}
+
+TEST_CASE("BeatsNumericConverterFormatter", "")
+{
+   MockedPrefs mockedPrefs;
+   MockedAudio mockedAudio;
```

Comment:
Are these necessary before a project can be created?  It is not obvious why so, but I am guessing the construction of the things hooked onto `ClientData::Site` by the libraries that lib-numeric-formats uses, made it necessary.  Those include lib-project-rate now.

Reply:
ProjectRate has a very complex initialization.

Reply:
That's right, and `QualitySettings::DefaultSampleRate` unusually uses a computed default value, using `AudioIOBase`

If it were just the constant 44100.0, I wonder if the test could simplify

Reply:
MockedAudio can be omitted if two assertions are removed from the AudioIO. It is not possible to omit MockedPrefs, because ProjectRate flushes unconditionally. 

It seems that some of this class are hanging on CI for some reason. I'm trying to figure that out, it is not reproducible locally or in the Windows Sandbox.